docs/performance/optimization.rst

```diff
@@ -0,0 +1,598 @@
+
+If you are using CASE expression for “formatting” see the previous point about
+formatting output as late as possible,
+
+but if you are using a CASE expression as part of a filter of other operation
+consider replacing it with an equivalent expression, for instance:
+
+.. code:: sql
+
+   SELECT SUM(a) as count_greater_than_10,...
+   FROM (
+     SELECT CASE WHEN field1 > 10 THEN 1 ELSE 0 END
+           , ...
+     FROM mytable
+     ...
+   ) subquery
+   ...;
+
+can be rewritten as
+
+.. code:: sql
+
+   SELECT COUNT(field1) FILTER (WHERE field1 > 10) as count_greater_than_10
+   FROM mytable;
+
+And
+
+.. code:: postgresql
+
+   SELECT *
+   FROM mytable
+   WHERE
+     CASE
+           WHEN $1 = 'ALL COUNTRIES' THEN true
+           WHEN $1 = mytable.country AND $2 = 'ALL CITIES' THEN true
+           ELSE $1 = mytable.country AND $2 = mytable.city
+     END;
+
+can be rewritten as
+
+.. code:: postgresql
+
+   SELECT *
+   FROM mytable
+   WHERE ($1 = 'ALL COUNTRIES')
+   OR ($1 = mytable.country AND $2 = 'ALL CITIES')
+   OR ($1 = mytable.country AND $2 = mytable.city)
+
+(the exact replacement expressions of course depend on the semantics of each
+case)
+
+.. _groups-instead-distinct:
+
+Use groupings instead of DISTINCT
+=================================
+
+(Reference: `Issue 13818`_)
+
+.. code:: sql
+
+   SELECT DISTINCT country FROM customers;
+
+use
+
+.. code:: sql
+
+   SELECT country FROM customers GROUP BY country;
+
+and instead of
+
+.. code:: sql
+
+   SELECT COUNT(DISTINCT a) FROM t;
+
+use
+
+.. code:: sql
+
+   SELECT COUNT(a)
+   FROM (
+           SELECT a
+           FROM t
+           GROUP BY a
+           ) tmp;
+
+.. _subqueries-instead-groups:
+
+Use subqueries instead of GROUP BY if the groups are already known
+==================================================================
+
+Consider the following query:
+
+.. code:: sql
+
+   SELECT customerid, SUM(order_amount) AS total
+   FROM customer_orders
+   GROUP BY customerid;
+
+This looks simple but to execute it CrateDB needs to keep the full result set in
+memory for all groups.
+
+If we already know what the groups will be we can use correlated subqueries
+instead:
+
+.. code:: sql
+
+   SELECT customerid,
+     (SELECT SUM(order_amount)
+      FROM customer_orders
+      WHERE customer_orders.customerid = customers.customerid
+     ) AS total
+   FROM customers;
+
+.. _group-large-and-complex-queries:
+
+************************************
+ Handling Large and Complex Queries
+************************************
+
+This section discusses strategies for breaking down complex operations on large
+datasets into manageable steps.
+
+.. _batch-operations:
+
+Batch operations
+================
+
+If you need to perform lots of UPDATEs or expensive INSERTs from SELECT, instead
+of doing them all in one go, adopt a batch approach where the operations are
+done on groups of records each time.
```

Comment:
What is the reason to do this? CrateDB uses an internal batching and throttling logic to run this in the most efficient way. Is there an issue with this?

Reply:
Maybe it should better read/say "[...] instead of invoking individual queries, please consider adopting a batch approach [...]"?